src/main/java/com/indeed/jiraactions/api/ApiCaller.java

```diff
@@ -3,178 +3,120 @@
-                    while (line != null) {
-                        sb.append(line).append(System.lineSeparator());
-                        line = errorBr.readLine();
-                    }
-                    errorBr.close();
-                } else {
-                    sb.append("Unable to open error stream for reading");
-                }
-                sb.append("\"");
+                sb.append("\"Code\": ").append(response.code()).append(",");
+                sb.append("\"Message\": \"").append(response.message()).append("\",");
+                sb.append("\"Error Body\": \"").append(responseBody.string()).append("\"");
+                sb.append('}');
+                sb.append('}');
+                throw new IOException(String.valueOf(response));
             }
-
-            sb.append("}");
-            sb.append("}");
-            log.error("Encountered connection error: " + sb);
+            return objectMapper.readTree(responseBody.string());
+        } catch (IOException e) {
             throw e;
-        } finally {
-            if (br != null) {
-                try {
-                    br.close();
-                } catch (final IOException ignored) {
-                }
-            }
         }
     }
 
-    private HttpsURLConnection getURLConnection(final String urlString) throws IOException {
-        final URL url = new URL(urlString);
-        final HttpsURLConnection urlConnection = (HttpsURLConnection) url.openConnection();
-        urlConnection.setRequestProperty("Authorization", authentication);
-
-        if(cookies.length() > 0) {
-            urlConnection.setRequestProperty("Cookie", cookies);
-        }
-        return urlConnection;
-    }
-
-    private void setCookies() {
-        final StringBuilder sb = new StringBuilder();
-        if(jsessionId != null) {
-            sb.append("JSESSIONID=").append(jsessionId);
-        }
-        if(upstream != null) {
-            if(sb.length() > 0) {
-                sb.append(";");
-            }
-            sb.append("upstream=").append(upstream);
-        }
-        cookies = sb.toString();
-    }
-
     private String getBasicAuth() {
         final String userPass = config.getJiraUsername() + ":" + config.getJiraPassword();
         final String basicAuth = "Basic " + new String(new Base64().encode(userPass.getBytes()));
         return basicAuth;
     }
+
+    private static CookieJar setupCookieJar() {
+        return new CookieJar() {
+            private final HashMap<String, List<Cookie>> cookieStore = new HashMap<>();
+
+            @Override
+            public void saveFromResponse(HttpUrl url, List<Cookie> cookies) {
+                cookieStore.put(url.host(), cookies);
+            }
+
+            @Override
+            public List<Cookie> loadForRequest(HttpUrl url) {
+                List<Cookie> cookies = cookieStore.get(url.host());
+                return cookies != null ? cookies : new ArrayList<Cookie>();
```

Comment:
Java 1.8 actually introduces a really cool function to do this, `computeIfAbsent()`. Something like `cookies.computeIfAbsent(cookies, () -> new ArrayList<Cookie>());` Prefer `computeIfAbsent()` to `getOrDefault()` because `cookies.getOrDefault(cookies, new ArrayList<Cookie());` would instantiate a new `ArrayList` every time, even if the map contains the value. The `Supplier<>` provides extra indirection that prevents this in `computeIfAbsent()`.

(I guess technically it would be even slightly more efficient to refactor the Supplier into a private variable.)